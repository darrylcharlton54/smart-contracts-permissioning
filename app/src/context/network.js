// Libs
import React, {
    useContext,
    useEffect,
    useState,
    createContext,
    useMemo
} from "react";
import { Drizzle, generateStore } from "drizzle";
import { drizzleReactHooks } from "drizzle-react";
// Constants
import drizzleOptions from "../drizzleOptions";
import Rules from "../contracts/Rules.json";
import Admin from "../contracts/Admin.json";
// Utils
import { getAllowedNetworks } from "../util/contracts";

const drizzleStore = generateStore(drizzleOptions);
const drizzle = new Drizzle(drizzleOptions, drizzleStore);

const NetworkContext = createContext();

/**
 * Provider for the network context that contains informations about the
 * blockchain provider
 * @param {Object} props Props given to the NetworkProvider
 * @return The provider with the following value:
 *  - isCorrectNetwork: true if the detected network is one of the allowed
 *  networks by the contracts, false if it is not, null if there is no detected
 *  network
 *  - setIsCorrectNetwork: setter of isCorrectNetwork
 *  - web3Initialized: true if Drizzle has initialized web3, false otherwise
 *  - setWeb3Initialized: setter for web3Initialized
 */
export const NetworkProvider = props => {
    const [isCorrectNetwork, setIsCorrectNetwork] = useState(null);
    const [web3Initialized, setWeb3Initialized] = useState(false);

    const value = useMemo(
        () => ({
            isCorrectNetwork,
            setIsCorrectNetwork,
            web3Initialized,
            setWeb3Initialized
        }),
        [
            isCorrectNetwork,
            setIsCorrectNetwork,
            web3Initialized,
            setWeb3Initialized
        ]
    );

    return (
        <drizzleReactHooks.DrizzleProvider drizzle={drizzle}>
            <NetworkContext.Provider value={value} {...props} />
        </drizzleReactHooks.DrizzleProvider>
    );
};

/**
 * Synchronize with the blockchain network
 * @return {Object} The network informations:
 *  - isCorrectNetwork: true if the detected network is one of the allowed
 *  networks by the contracts, false if it is not, null if there is no detected
 *  network,
 *  - networkId: The id of the network,
 *  - web3Initialized: true if Drizzle has initialized web3, false otherwise
 */
export const useNetwork = () => {
    const context = useContext(NetworkContext);
    if (!context) {
        throw new Error("useData must be used within a DataProvider.");
    }

    const {
        isCorrectNetwork,
        setIsCorrectNetwork,
        web3Initialized,
        setWeb3Initialized
    } = context;

    const { networkId, status } = drizzleReactHooks.useDrizzleState(
        drizzleState => ({
            status: drizzleState.web3.status,
            networkId: drizzleState.web3.networkId
        })
    );

    useEffect(() => {
        if (status === "initialized") {
            const allowedNetworks = getAllowedNetworks([Rules, Admin]);
            if (networkId) {
                const isCorrectNetwork = allowedNetworks.includes(networkId);
                setIsCorrectNetwork(isCorrectNetwork);
            }
            setWeb3Initialized(true);
        }
    }, [networkId, setIsCorrectNetwork, status, setWeb3Initialized]);

    return {
        isCorrectNetwork,
        networkId,
        web3Initialized
    };
};
