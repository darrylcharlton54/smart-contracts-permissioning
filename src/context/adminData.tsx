// Libs
import React, { createContext, useContext, useMemo, useState, useEffect } from "react";
import { drizzleReactHooks } from "drizzle-react";

type ContextType = {
    admins?: string[],
    setAdmins?: (admins: string[]) => void
}

const AdminDataContext = createContext<ContextType>({})

/**
 * Provider for the data context that contains the Admin whitelist
 * @param {Object} props Props given to the AdminDataProvider
 * @return The provider with the following value:
 *  - admins: list of Admin accounts from Admin Rules contract
 *  - setAdmins: setter for the Admin list state
 */
export const AdminDataProvider: React.FC = (props: React.Props<{}>) => {
    const [admins, setAdmins] = useState<string[]>([])
    const value = useMemo(() => ({ admins, setAdmins }), [
        admins,
        setAdmins
    ])
    return <AdminDataContext.Provider value={value} {...props} />
}

/**
 * Fetch the appropriate Admin data on chain and synchronize with it
 * @return {Object} Contains data of interest:
  *  - dataReady: true if Admin whitelist has been correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isAdmin: user in an Admin,
 *  - whitelist: list of whitelist nodes from Node contract,
 */
export const useAdminData = () => {
    const context = useContext(AdminDataContext)

    if (!context) {
        throw new Error("useAdminData must be used within an AdminDataProvider.")
    }

    const { admins, setAdmins } = context

    const { userAddress } = drizzleReactHooks.useDrizzleState((drizzleState: any) => ({
        userAddress: drizzleState.accounts[0]
    }));

    const { useCacheCall } = drizzleReactHooks.useDrizzle();
    const adminList: string[] = useCacheCall("Admin", "getAdmins");

    useEffect(() => {
        setAdmins!(adminList || [])
        }, [adminList, setAdmins]
    )


    const dataReady = useMemo(
        () =>
            Array.isArray(admins),
            [admins]
    );

    const isAdmin = useMemo(
        () => (dataReady && admins ? admins.includes(userAddress) : false),
        [dataReady, admins, userAddress]
    );

    const formattedAdmins = useMemo(() => {
        return admins
            ? admins
                  .map(address => ({
                      address,
                      identifier: address.toLowerCase(),
                      status: "active"
                  }))
                  .reverse()
            : undefined;
    }, [admins]);

    return {
        dataReady,
        userAddress,
        isAdmin,
        admins: formattedAdmins
    }
}
