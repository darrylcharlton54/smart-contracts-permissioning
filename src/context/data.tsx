// Libs
import React, {
    createContext,
    useContext,
    useEffect,
    useState,
    useMemo
} from "react";
import { drizzleReactHooks } from "drizzle-react";

// Utils
import { paramsToIdentifier } from "../util/enodetools";

type Enode = {enodeHigh: string, enodeLow: string, identifier: string, ip: string, port: string};
type Account = {address: string};

type ContextType = {
    nodeWhitelist?: Enode[],
    setNodeWhitelist?: (enode: Enode[]) => void,
    accountWhitelist?: Account[],
    setAccountWhitelist?: (account: Account[]) => void
}

const DataContext = createContext<ContextType>({});

/**
 * Provider for the data context that contains the whitelist
 * @param {Object} props Props given to the DataProvider
 * @return The provider with the following value:
 *  - nodeWhitelist: list of whiteliist enode from Node Rules contract
 *  - setNodeWhitelist: setter for the whitelist state
 */
export const DataProvider: React.FC = (props: React.Props<{}>) => {
    const [nodeWhitelist, setNodeWhitelist] = useState<Enode[]>([]);
    const [accountWhitelist, setAccountWhitelist] = useState<Account[]>([]);

    const value = useMemo(() => ({ nodeWhitelist, setNodeWhitelist, accountWhitelist, setAccountWhitelist }), [
        nodeWhitelist,
        setNodeWhitelist,
        accountWhitelist,
        setAccountWhitelist
    ]);
    return <DataContext.Provider value={value} {...props} />;
};

/**
 * Fetch the appropriate data on chain and synchronize with it
 * @return {Object} Contains data of interest:
 *  - admins: list of admin address from Admin contract,
 *  - dataReady: true if isReadOnly, whitelist and admins are correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isAdmin: true if address of the user is includes in the admin list
 *  - node: Object containing node relevant data
 *    - isReadOnly: Rules contract is lock or unlock,
 *    - whitelist: list of whitelist enode from Rules contract,
 */
export const useData = () => {
    const context = useContext(DataContext);
    if (!context) {
        throw new Error("useData must be used within a DataProvider.");
    }

    const { nodeWhitelist, setNodeWhitelist, accountWhitelist, setAccountWhitelist } = context;

    const { userAddress } = drizzleReactHooks.useDrizzleState((drizzleState: any) => ({
        userAddress: drizzleState.accounts[0]
    }));

    const { drizzle, useCacheCall } = drizzleReactHooks.useDrizzle();

    const nodeIsReadOnly: boolean = useCacheCall("NodeRules", "isReadOnly");
    const nodeWhitelistSize: number = useCacheCall("NodeRules", "getSize");
    const accountIsReadOnly: boolean = useCacheCall("AccountRules", "isReadOnly");
    const accountWhitelistSize: number = useCacheCall("AccountRules", "getSize");
    const admins: string[] = useCacheCall("Admin", "getAdmins");

    const { getByIndex: getNodeByIndex } = drizzle.contracts.NodeRules.methods;

    useEffect(() => {
        const promises = [];
        for (let index = 0; index < nodeWhitelistSize; index++) {
            promises.push(getNodeByIndex(index).call());
        }
        Promise.all(promises).then(responses => {
            const updatedNodeWhitelist = responses.map(
                ({ enodeHigh, enodeLow, ip, port }) => ({
                    enodeHigh,
                    enodeLow,
                    ip,
                    port,
                    identifier: paramsToIdentifier({
                        enodeHigh,
                        enodeLow,
                        ip,
                        port
                    })
                })
            );
            setNodeWhitelist!(updatedNodeWhitelist);
        });
    }, [nodeWhitelistSize, setNodeWhitelist, getNodeByIndex]);

    const { getByIndex: getAccountByIndex } = drizzle.contracts.AccountRules.methods;

    useEffect(() => {
        const promises = [];
        for (let index = 0; index < accountWhitelistSize; index++) {
            promises.push(getAccountByIndex(index).call());
        }
        Promise.all(promises).then(responses => {
            const updatedAccountWhitelist = responses.map((address: string) => ({address}));
            setAccountWhitelist!(updatedAccountWhitelist);
        });
    }, [accountWhitelistSize, setAccountWhitelist, getAccountByIndex]);

    const dataReady = useMemo(
        () =>
            typeof nodeIsReadOnly === "boolean" &&
            Array.isArray(admins) &&
            Array.isArray(nodeWhitelist) &&
            Array.isArray(accountWhitelist),
        [nodeIsReadOnly, admins, nodeWhitelist, accountWhitelist]
    );

    const isAdmin = useMemo(
        () => (dataReady ? admins.includes(userAddress) : false),
        [dataReady, admins, userAddress]
    );

    const formattedAdmins = useMemo(() => {
        return admins
            ? admins
                  .map(address => ({
                      address,
                      identifier: address,
                      status: "active"
                  }))
                  .reverse()
            : undefined;
    }, [admins]);

    const formattedNodeWhitelist = useMemo(() => {
        return nodeWhitelist
            ? nodeWhitelist
                  .map(enode => ({ ...enode, status: "active" }))
                  .reverse()
            : undefined;
    }, [nodeWhitelist]);

    const formattedAccountWhitelist = useMemo(() => {
        return accountWhitelist
            ? accountWhitelist
                  .map(account => ({ ...account, status: "active" }))
                  .reverse()
            : undefined;
    }, [accountWhitelist]);

    return {
        userAddress,
        dataReady,
        isAdmin,
        admins: formattedAdmins,
        node: {
            whitelist: formattedNodeWhitelist,
            isReadOnly: nodeIsReadOnly
        },
        account: {
            whitelist: formattedAccountWhitelist,
            isReadOnly: accountIsReadOnly
        }
    };
};
