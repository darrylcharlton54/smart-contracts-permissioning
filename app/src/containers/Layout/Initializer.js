// Libs
import React from "react";
import PropTypes from "prop-types";
// Context
import { useNetwork } from "../../context/network";

const Initializer = ({ children, NoProvider, WrongNetwork }) => {
    const { networkId, isCorrectNetwork, web3Initialized } = useNetwork();
    const noDetectedProvider = networkId === undefined;
    return !web3Initialized ? null : noDetectedProvider ? (
        <NoProvider />
    ) : isCorrectNetwork ? (
        children
    ) : (
        <WrongNetwork networkId={networkId} />
    );
};

Initializer.propTypes = {
    children: PropTypes.object.isRequired,
    NoProvider: PropTypes.object.isRequired,
    WrongNetwork: PropTypes.object.isRequired
};

export default Initializer;
