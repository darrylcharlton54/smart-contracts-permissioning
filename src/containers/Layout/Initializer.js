// Libs
import React from 'react';
import PropTypes from 'prop-types';
// Context
import { useNetwork } from '../../context/network';

const Initializer = ({ children, NoProvider, WrongNetwork }) => {
  const { networkId, isCorrectNetwork, web3Initialized } = useNetwork();
  const noDetectedProvider = networkId === undefined;
  return !web3Initialized ? null : noDetectedProvider ? (
    <NoProvider />
  ) : isCorrectNetwork ? (
    children
  ) : (
    <WrongNetwork networkId={networkId} />
  );
};

Initializer.propTypes = {
  children: PropTypes.object.isRequired,
  NoProvider: PropTypes.func.isRequired,
  WrongNetwork: PropTypes.func.isRequired
};

export default Initializer;
