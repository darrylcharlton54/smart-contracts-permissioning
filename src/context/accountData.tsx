import React, { createContext, useContext, useEffect, useState, useMemo } from 'react';
import { AccountRules } from '../chain/@types/AccountRules';
import { accountRulesFactory } from '../chain/contracts/AccountRules';
import { useNetwork } from './network';

type Account = { address: string };

type ContextType =
  | {
      accountWhitelist: Account[];
      setAccountWhitelist: React.Dispatch<React.SetStateAction<Account[]>>;
      accountReadOnly?: boolean;
      setAccountReadOnly: React.Dispatch<React.SetStateAction<boolean | undefined>>;
      accountRulesContract?: AccountRules;
      setAccountRulesContract: React.Dispatch<React.SetStateAction<AccountRules | undefined>>;
    }
  | undefined;

const AccountDataContext = createContext<ContextType>(undefined);

const loadAccountData = (
  accountRulesContract: AccountRules | undefined,
  setAccountWhitelist: (account: Account[]) => void,
  setAccountReadOnly: (readOnly?: boolean) => void
) => {
  if (accountRulesContract === undefined) {
    setAccountWhitelist([]);
    setAccountReadOnly(undefined);
  } else {
    accountRulesContract.functions.isReadOnly().then(isReadOnly => setAccountReadOnly(isReadOnly));
    accountRulesContract.functions.getSize().then(whitelistSize => {
      const whitelistElementsPromises = [];
      for (let i = 0; whitelistSize.gt(i); i++) {
        whitelistElementsPromises.push(accountRulesContract.functions.getByIndex(i));
      }
      Promise.all(whitelistElementsPromises).then(responses => {
        setAccountWhitelist(responses.map(address => ({ address })));
      });
    });
  }
};

/**
 * Provider for the data context that contains the account whitelist
 * @param {Object} props Props given to the AccountDataProvider
 * @return The provider with the following value:
 *  - accountWhitelist: list of whiteliist accounts from Account Rules contract
 *  - setNodeWhitelist: setter for the whitelist state
 */
export const AccountDataProvider: React.FC<{}> = props => {
  const [accountWhitelist, setAccountWhitelist] = useState<Account[]>([]);
  const [accountReadOnly, setAccountReadOnly] = useState<boolean | undefined>(undefined);
  const [accountRulesContract, setAccountRulesContract] = useState<AccountRules | undefined>(undefined);

  const value = useMemo(
    () => ({
      accountWhitelist,
      setAccountWhitelist,
      accountReadOnly,
      setAccountReadOnly,
      accountRulesContract,
      setAccountRulesContract
    }),
    [
      accountWhitelist,
      setAccountWhitelist,
      accountReadOnly,
      setAccountReadOnly,
      accountRulesContract,
      setAccountRulesContract
    ]
  );

  const { accountIngressContract } = useNetwork();

  useEffect(() => {
    if (accountIngressContract === undefined) {
      setAccountRulesContract(undefined);
    } else {
      accountRulesFactory(accountIngressContract).then(contract => {
        setAccountRulesContract(contract);
        contract.removeAllListeners('AccountAdded');
        contract.removeAllListeners('AccountRemoved');
        contract.on('AccountAdded', () => {
          loadAccountData(contract, setAccountWhitelist, setAccountReadOnly);
        });
        contract.on('AccountRemoved', () => {
          loadAccountData(contract, setAccountWhitelist, setAccountReadOnly);
        });
      });
    }
  }, [accountIngressContract, setAccountWhitelist, setAccountReadOnly]);

  return <AccountDataContext.Provider value={value} {...props} />;
};

/**
 * Fetch the appropriate account data on chain and synchronize with it
 * @return {Object} Contains data of interest:
 *  - dataReady: true if isReadOnly and account whitelist are correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isReadOnly: Account contract is lock or unlock,
 *  - whitelist: list of whitelist accounts from Account contract,
 */
export const useAccountData = () => {
  const context = useContext(AccountDataContext);

  if (!context) {
    throw new Error('useAccountData must be used within an AccountDataProvider.');
  }

  const { accountWhitelist, setAccountWhitelist, accountReadOnly, setAccountReadOnly, accountRulesContract } = context;

  useEffect(() => {
    loadAccountData(accountRulesContract, setAccountWhitelist, setAccountReadOnly);
  }, [accountRulesContract, setAccountWhitelist, setAccountReadOnly]);

  const formattedAccountWhitelist = useMemo(() => {
    return accountWhitelist
      .map(account => ({
        ...account,
        identifier: account.address.toLowerCase(),
        status: 'active'
      }))
      .reverse();
  }, [accountWhitelist]);

  const dataReady = useMemo(() => {
    return accountRulesContract !== undefined && accountReadOnly !== undefined && accountWhitelist !== undefined;
  }, [accountRulesContract, accountReadOnly, accountWhitelist]);

  return {
    dataReady,
    whitelist: formattedAccountWhitelist,
    isReadOnly: accountReadOnly,
    accountRulesContract
  };
};
