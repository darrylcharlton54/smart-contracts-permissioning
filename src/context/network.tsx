// Libs
import React, { useContext, useEffect, useState, createContext, useMemo } from 'react';
import { Drizzle, generateStore } from 'drizzle';
import { drizzleReactHooks } from 'drizzle-react';
// Constants
import drizzleOptions from '../drizzleOptions';
import NodeRules from '../chain/abis/NodeRules.json';
import AccountRules from '../chain/abis/AccountRules.json';
import Admin from '../chain/abis/Admin.json';
// Utils
import { getAllowedNetworks } from '../util/contracts';
import { useConfig } from '../context/configData';

import { providerFactory } from '../chain/provider';
import { AccountIngress } from '../chain/@types/AccountIngress';
import { accountIngressFactory } from '../chain/contracts/AccountIngress';
import { NodeIngress } from '../chain/@types/NodeIngress';
import { nodeIngressFactory } from '../chain/contracts/NodeIngress';

type ContextType =
  | {
      isCorrectNetwork?: boolean;
      setIsCorrectNetwork: React.Dispatch<React.SetStateAction<boolean | undefined>>;
      web3Initialized: boolean;
      setWeb3Initialized: React.Dispatch<React.SetStateAction<boolean>>;
      contracts: {
        accountIngressContract?: AccountIngress;
        setAccountIngressContract: React.Dispatch<React.SetStateAction<AccountIngress | undefined>>;
        nodeIngressContract?: NodeIngress;
        setNodeIngressContract: React.Dispatch<React.SetStateAction<NodeIngress | undefined>>;
      };
    }
  | undefined;

const NetworkContext = createContext<ContextType>(undefined);

/**
 * Provider for the network context that contains informations about the
 * blockchain provider
 * @param {Object} props Props given to the NetworkProvider
 * @return The provider with the following value:
 *  - isCorrectNetwork: true if the detected network is one of the allowed
 *  networks by the contracts, false if it is not, null if there is no detected
 *  network
 *  - setIsCorrectNetwork: setter of isCorrectNetwork
 *  - web3Initialized: true if Drizzle has initialized web3, false otherwise
 *  - setWeb3Initialized: setter for web3Initialized
 */
export const NetworkProvider: React.FC<{}> = props => {
  const [isCorrectNetwork, setIsCorrectNetwork] = useState<boolean | undefined>(undefined);
  const [web3Initialized, setWeb3Initialized] = useState<boolean>(false);
  const [accountIngressContract, setAccountIngressContract] = useState<AccountIngress | undefined>(undefined);
  const [nodeIngressContract, setNodeIngressContract] = useState<NodeIngress | undefined>(undefined);

  const config = useConfig();
  const [drizzle] = useState<Drizzle>(() => {
    const options = drizzleOptions(config);

    const drizzleStore = generateStore(options);
    return new Drizzle(options, drizzleStore);
  });

  useEffect(() => {
    providerFactory().then(provider => {
      accountIngressFactory(config, provider).then(accountIngress => setAccountIngressContract(accountIngress));
      nodeIngressFactory(config, provider).then(nodeIngress => setNodeIngressContract(nodeIngress));
    });
  }, [config]);

  const value = useMemo(
    () => ({
      isCorrectNetwork,
      setIsCorrectNetwork,
      web3Initialized,
      setWeb3Initialized,
      contracts: {
        accountIngressContract,
        setAccountIngressContract,
        nodeIngressContract,
        setNodeIngressContract
      }
    }),
    [
      isCorrectNetwork,
      setIsCorrectNetwork,
      web3Initialized,
      setWeb3Initialized,
      accountIngressContract,
      setAccountIngressContract,
      nodeIngressContract,
      setNodeIngressContract
    ]
  );

  return (
    <drizzleReactHooks.DrizzleProvider drizzle={drizzle}>
      <NetworkContext.Provider value={value} {...props} />
    </drizzleReactHooks.DrizzleProvider>
  );
};

/**
 * Synchronize with the blockchain network
 * @return {Object} The network informations:
 *  - isCorrectNetwork: true if the detected network is one of the allowed
 *  networks by the contracts, false if it is not, null if there is no detected
 *  network,
 *  - networkId: The id of the network,
 *  - web3Initialized: true if Drizzle has initialized web3, false otherwise
 */
export const useNetwork = () => {
  const context = useContext(NetworkContext);
  if (!context) {
    throw new Error('useNetwork must be used within a DataProvider.');
  }

  const { isCorrectNetwork, setIsCorrectNetwork, web3Initialized, setWeb3Initialized, contracts } = context;

  const { networkId, status } = drizzleReactHooks.useDrizzleState((drizzleState: any) => ({
    status: drizzleState.web3.status,
    networkId: drizzleState.web3.networkId
  }));

  useEffect(() => {
    if (status === 'initialized') {
      const allowedNetworks = getAllowedNetworks([AccountRules, NodeRules, Admin]);
      if (networkId) {
        const isCorrectNetwork = allowedNetworks.includes(networkId);
        setIsCorrectNetwork(isCorrectNetwork);
      }
      setWeb3Initialized(true);
    }
  }, [networkId, setIsCorrectNetwork, status, setWeb3Initialized]);

  return {
    isCorrectNetwork,
    networkId,
    web3Initialized,
    accountIngressContract: contracts.accountIngressContract,
    nodeIngressContract: contracts.nodeIngressContract
  };
};
