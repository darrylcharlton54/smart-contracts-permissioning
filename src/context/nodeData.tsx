// Libs
import React, { createContext, useContext, useEffect, useState, useMemo } from 'react';
import { NodeRules } from '../chain/@types/NodeRules';
import { nodeRulesFactory } from '../chain/contracts/NodeRules';
import { useNetwork } from './network';

// Utils
import { paramsToIdentifier, Enode as RawEnode } from '../util/enodetools';

type Enode = RawEnode & { identifier: string };

type ContextType =
  | {
      nodeWhitelist: Enode[];
      setNodeWhitelist: React.Dispatch<React.SetStateAction<Enode[]>>;
      nodeReadOnly?: boolean;
      setNodeReadOnly: React.Dispatch<React.SetStateAction<boolean | undefined>>;
      nodeRulesContract?: NodeRules;
      setNodeRulesContract: React.Dispatch<React.SetStateAction<NodeRules | undefined>>;
    }
  | undefined;

const DataContext = createContext<ContextType>(undefined);

const loadNodeData = (
  nodeRulesContract: NodeRules | undefined,
  setNodeWhitelist: (account: Enode[]) => void,
  setNodeReadOnly: (readOnly?: boolean) => void
) => {
  if (nodeRulesContract === undefined) {
    setNodeWhitelist([]);
    setNodeReadOnly(undefined);
  } else {
    nodeRulesContract.functions.isReadOnly().then(isReadOnly => setNodeReadOnly(isReadOnly));
    nodeRulesContract.functions.getSize().then(whitelistSize => {
      const whitelistElementPromises = [];
      for (let i = 0; whitelistSize.gt(i); i++) {
        whitelistElementPromises.push(nodeRulesContract.functions.getByIndex(i));
      }
      Promise.all(whitelistElementPromises).then(responses => {
        const updatedNodeWhitelist = responses.map(r => {
          const withStringyPort = { ...r, port: r.port.toString() };
          return {
            ...withStringyPort,
            identifier: paramsToIdentifier(withStringyPort)
          };
        });
        setNodeWhitelist(updatedNodeWhitelist);
      });
    });
  }
};

/**
 * Provider for the data context that contains the node whitelist
 * @param {Object} props Props given to the NodeDataProvider
 * @return The provider with the following value:
 *  - nodeWhitelist: list of whiteliist enode from Node Rules contract
 *  - setNodeWhitelist: setter for the whitelist state
 */
export const NodeDataProvider: React.FC<{}> = props => {
  const [nodeWhitelist, setNodeWhitelist] = useState<Enode[]>([]);
  const [nodeReadOnly, setNodeReadOnly] = useState<boolean | undefined>(undefined);
  const [nodeRulesContract, setNodeRulesContract] = useState<NodeRules | undefined>(undefined);

  const value = useMemo(
    () => ({ nodeWhitelist, setNodeWhitelist, nodeReadOnly, setNodeReadOnly, nodeRulesContract, setNodeRulesContract }),
    [nodeWhitelist, setNodeWhitelist, nodeReadOnly, setNodeReadOnly, nodeRulesContract, setNodeRulesContract]
  );

  const { nodeIngressContract } = useNetwork();

  useEffect(() => {
    if (nodeIngressContract === undefined) {
      setNodeRulesContract(undefined);
    } else {
      nodeRulesFactory(nodeIngressContract).then(contract => {
        setNodeRulesContract(contract);
        contract.removeAllListeners('NodeAdded');
        contract.removeAllListeners('NodeRemoved');
        contract.on('NodeAdded', () => {
          loadNodeData(contract, setNodeWhitelist, setNodeReadOnly);
        });
        contract.on('NodeRemoved', () => {
          loadNodeData(contract, setNodeWhitelist, setNodeReadOnly);
        });
      });
    }
  }, [nodeIngressContract]);

  return <DataContext.Provider value={value} {...props} />;
};

/**
 * Fetch the appropriate node data on chain and synchronize with it
 * @return {Object} Contains data of interest:
 *  - dataReady: true if isReadOnly and node whitelist are correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isReadOnly: Node contract is lock or unlock,
 *  - whitelist: list of whitelist nodes from Node contract,
 */
export const useNodeData = () => {
  const context = useContext(DataContext);
  if (!context) {
    throw new Error('useNodeData must be used within a NodeDataProvider.');
  }

  const { nodeWhitelist, setNodeWhitelist, nodeReadOnly, setNodeReadOnly, nodeRulesContract } = context;

  useEffect(() => {
    loadNodeData(nodeRulesContract, setNodeWhitelist, setNodeReadOnly);
  }, [nodeRulesContract, setNodeWhitelist, setNodeReadOnly]);

  const formattedNodeWhitelist = useMemo(() => {
    return nodeWhitelist.map(enode => ({ ...enode, status: 'active' })).reverse();
  }, [nodeWhitelist]);

  const dataReady = useMemo(() => {
    return nodeRulesContract !== undefined && nodeReadOnly !== undefined && nodeWhitelist !== undefined;
  }, [nodeRulesContract, nodeReadOnly, nodeWhitelist]);

  return {
    dataReady,
    whitelist: formattedNodeWhitelist,
    isReadOnly: nodeReadOnly,
    nodeRulesContract
  };
};
