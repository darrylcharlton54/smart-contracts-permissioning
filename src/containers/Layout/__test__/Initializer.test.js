// Libs
import React from "react";
import toJson from "enzyme-to-json";
import { mount } from "enzyme";
// Components
import Initializer from "../Initializer";

import { useNetwork } from "../../../context/network";

jest.mock("../../../context/network", () => {
    return {
        useNetwork: jest.fn()
    };
});

const NoProvider = () => <div className="noProvider" />;
const WrongNetwork = ({ networkId }) => <div className="wrongNetwork" />;

describe("<Initializer />", () => {
    let wrapper;

    beforeAll(() => {
        useNetwork.mockImplementation(() => ({
            web3Initialized: false,
            networkId: undefined,
            isCorrectNetwork: null
        }));
    });

    beforeEach(() => {
        jest.clearAllMocks();

        wrapper = mount(
            <Initializer NoProvider={NoProvider} WrongNetwork={WrongNetwork}>
                <div className="test" />
            </Initializer>
        );
    });

    it("has called useNetwork once", () => {
        expect(useNetwork).toHaveBeenCalledTimes(1);
    });

    it("contains NoProvider={NoProvider} as a props", () => {
        expect(wrapper.props().NoProvider).toEqual(NoProvider);
    });

    it("contains WrongNetwork={WrongNetwork} as a props", () => {
        expect(wrapper.props().WrongNetwork).toEqual(WrongNetwork);
    });

    it('contains children={<div className="test" />} as a props', () => {
        expect(wrapper.props().children).toEqual(<div className="test" />);
    });

    describe("web3Initialized=false", () => {
        beforeAll(() => {
            useNetwork.mockImplementation(() => ({
                web3Initialized: false,
                networkId: undefined,
                isCorrectNetwork: null
            }));
        });

        it("does not render children when passed in", () => {
            expect(wrapper.contains(<div className="test" />)).toEqual(false);
        });

        it("contains no children", () => {
            expect(wrapper.children()).toHaveLength(0);
        });

        it("matches snapshot", () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });

    describe("web3Initialized=true, networkId=undefined", () => {
        beforeAll(() => {
            useNetwork.mockImplementation(() => ({
                web3Initialized: true,
                networkId: undefined,
                isCorrectNetwork: null
            }));
        });

        it("does not render children when passed in", () => {
            expect(wrapper.contains(<div className="test" />)).toEqual(false);
        });

        it("contains one children", () => {
            expect(wrapper.children()).toHaveLength(1);
        });

        it("contains NoProvider", () => {
            expect(wrapper.contains(<NoProvider />)).toEqual(true);
        });

        it("matches snapshot", () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });

    describe("web3Initialized=true, networkId=0, isCorrectNetwork=false", () => {
        beforeAll(() => {
            useNetwork.mockImplementation(() => ({
                web3Initialized: true,
                networkId: 0,
                isCorrectNetwork: false
            }));
        });

        it("does not render children when passed in", () => {
            expect(wrapper.contains(<div className="test" />)).toEqual(false);
        });

        it("contains one children", () => {
            expect(wrapper.children()).toHaveLength(1);
        });

        it("contains NoProvider", () => {
            expect(wrapper.contains(<WrongNetwork networkId={0} />)).toEqual(
                true
            );
        });

        it("matches snapshot", () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });

    describe("web3Initialized=true, networkId=0, isCorrectNetwork=true", () => {
        beforeAll(() => {
            useNetwork.mockImplementation(() => ({
                web3Initialized: true,
                networkId: 0,
                isCorrectNetwork: true
            }));
        });

        it("render children when passed in", () => {
            expect(wrapper.contains(<div className="test" />)).toEqual(true);
        });

        it("matches snapshot", () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
    });
});
