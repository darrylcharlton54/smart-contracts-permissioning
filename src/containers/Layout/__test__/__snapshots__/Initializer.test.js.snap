// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Initializer /> web3Initialized=false matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
/>
`;

exports[`<Initializer /> web3Initialized=true, networkId=0, isCorrectNetwork=false matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
>
  <WrongNetwork
    networkId={0}
  >
    <div
      className="wrongNetwork"
    />
  </WrongNetwork>
</Initializer>
`;

exports[`<Initializer /> web3Initialized=true, networkId=0, isCorrectNetwork=true matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
>
  <div
    className="test"
  />
</Initializer>
`;

exports[`<Initializer /> web3Initialized=true, networkId=undefined matches snapshot 1`] = `
<Initializer
  NoProvider={[Function]}
  WrongNetwork={[Function]}
>
  <NoProvider>
    <div
      className="noProvider"
    />
  </NoProvider>
</Initializer>
`;
